### PR TITLE
Adding armature to computeallterms

### PR DESCRIPTION
**Pull request by proyan**
1) Remove rotorInertia and rotorGearRatio. Replace with armature.
2) add armature to compute all terms
3) test.

Currently it only adds the armature to the mass diagonal. In case you decide to add the coriolis and centrifugal effects, please reflect it in the computeallterms as well.

For now, this deals partially with what was discussed in #671

### CODE REVIEW

File: unittest/compute-all-terms.cpp [hunk 1]
```diff
@@ -162,4 +162,61 @@ BOOST_AUTO_TEST_CASE ( test_against_algo )
...
+
+  pinocchio::Model model; buildModels::humanoidRandom(model);
+  pinocchio::Data data(model); data.M.fill (0.);
+  pinocchio::Data data_other(model); data_other.M.fill (0.);
+
+  VectorXd q (VectorXd::Random(model.nq));
+  VectorXd v (VectorXd::Random(model.nv));
+
+  // ----Check that Original computeAllTerms still works-
+  q.setZero ();
```
Comment (jcarpent) [inline unittest/compute-all-terms.cpp]: Why q et v set to Zero?

File: unittest/compute-all-terms.cpp [hunk 2]
```diff
@@ -162,4 +162,61 @@ BOOST_AUTO_TEST_CASE ( test_against_algo )
...
+  potentialEnergy(model, data_other, q, true);
+
+  BOOST_CHECK (data.nle.isApprox(data_other.nle, 1e-12));
+  BOOST_CHECK (Eigen::MatrixXd(data.M.triangularView<Eigen::Upper>())
+              .isApprox(Eigen::MatrixXd(data_other.M.triangularView<Eigen::Upper>()), 1e-12));
+  BOOST_CHECK (data.J.isApprox(data_other.J, 1e-12));
+  BOOST_CHECK (data.Jcom.isApprox(data_other.Jcom, 1e-12));
+
+  // ----Check that computeAllTerms with armature works---
+  VectorXd armature(VectorXd::Random(model.nv));
```
Comment (jcarpent) [inline unittest/compute-all-terms.cpp]: armature are positive reals

File: src/algorithm/compute-all-terms.hpp [hunk 3]
```diff
@@ -233,6 +233,9 @@ namespace pinocchio
       Pass2::run(model.joints[i],data.joints[i],
                  typename Pass2::ArgsType(model,data));
     }
+
+    //Add the armature value. By default, this is a zero vector.
```
Comment (jcarpent) [inline src/algorithm/compute-all-terms.hpp]: You should add an assert here, or in the checking of the algo, to check if the armature is indeed positive.

**Comment by proyan** [comment]
By the way if you want to account for the effect on the nle, we should put back the rotorInertia and gearRatio inside Model class.

File: unittest/compute-all-terms.cpp (hunk 1 above)
Comment (proyan) [reply unittest/compute-all-terms.cpp]: This was a mistake. I will correct it.

**Comment by jcarpent** [comment]
> By the way if you want to account for the effect on the nle, we should put back the rotorInertia and gearRatio inside Model class.

I think the armature is sufficient (but maybe I'm wrong), as it corresponds to the handling of the rotor inertia as observed from the joint axis.

File: unittest/compute-all-terms.cpp (hunk 1 above)
Comment (jcarpent) [reply unittest/compute-all-terms.cpp]: Ok. Perfect.

**Comment by nmansard** [comment]
Thanks for the contribution. I answer first because we discussed that orally. @jcarpent please react after me. 
I like the change in the parsing and storage of armature elements. 
I am afraid of the issue raised by @gabrielebndn in #676: the dynamic algorithms should all be consistent, i.e. we should also propagate the armature to e.g. mass-inverse, kinematic energy, etc.

I suggest that some post-processing algorithms are added in a dedicated algorithm/armature.hpp file, that would minimally modify the existing algorithm and offer all classical functionalities with armature. Two set of algos can be separated:
- for rnea, crba, computeallterms, this is pure post-processing: run the classical algo, add the armature to the result, returns.
- for aba and massinv algo: as you raised it, you need to re-write a dedicated 3-rd pass of the algorithm, and can reuse the pass 1 and 2.

**Comment by nmansard** [comment]
For this PR, I suggest you only keep the modification of parser and model. This is a consensual one, it should be accepted quickly. In a second PR, please introduce a armature.hpp file with the post-processing of computeallterms. You will finally insert the ABA and Minv new algos in a later PR.

File: unittest/compute-all-terms.cpp (hunk 2 above)
Comment (proyan) [reply unittest/compute-all-terms.cpp]: Done.

File: unittest/compute-all-terms.cpp (hunk 1 above)
Comment (proyan) [reply unittest/compute-all-terms.cpp]: Done.

**Comment by jcarpent** [comment]
>     * for rnea, crba, computeallterms, this is pure post-processing: run the classical algo, add the armature to the result, returns.

For RNEA, I think we should take into account centrifugal and Coriolis effects too.

File: src/algorithm/compute-all-terms.hpp (hunk 3 above)
Comment (proyan) [reply src/algorithm/compute-all-terms.hpp]: Done.

**Comment by jcarpent** [comment]
I think we should try to avoid to put all the versions of armature algorithms in the same file. It will not help the maintenance and the readability, mostly if all the algorithms have to be double.

In terms of naming, I would suggest to add the namespace armature and keep the same naming for the algorithms.

@nmansard What do you think?

**Comment by jcarpent** [comment]
I don't know what to do with this PR. 
I can merge it because it is problematic as the changes are currently transparent for the end users.
Or, we can start to address the issue raised by @nmansard by providing separates signatures when taking into account the armature information.

@proyan @nmansard I think both of you are the main concerned by this question.

**Comment by proyan** [comment]
If we are okay with duplicating and adding separate wrappers for the algorithms, I think you should wait before merging this PR, and I can create a CATWithArmature in armature.hpp. Otherwise, you should merge.
However, I don't know what the resolution of #671 is, so I can't suggest the way forward.

(I personally would prefer the separate signatures)

**Comment by jcarpent** [comment]
In my case, I don't like the withArmature suffix. While the namespace seems more elegant, and easier to use by simply writing `using armature::algo` for instance.
In addition, it allows us to put directly the code inside the dedicated place, instead of adding a single file with a lot of signatures, not relevant if you want to use only `crba` for instance.

**Comment by proyan** [comment]
That is okay with me. I'll close this and make a separate PR which puts CAT it in a separate namespace.

**Comment by jcarpent** [comment]
Ok. Sounds great. Thanks a lot!

**Pull request closed without merging**